pkg/operator/vsphereproblemdetector/monitoring.go

```diff
@@ -0,0 +1,80 @@
+package vsphereproblemdetector
+
+import (
+	"context"
+	"time"
+
+	operatorapi "github.com/openshift/api/operator/v1"
+	"github.com/openshift/cluster-storage-operator/pkg/csoclients"
+	"github.com/openshift/cluster-storage-operator/pkg/generated"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceapply"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
+)
+
+type monitoringController struct {
+	operatorClient v1helpers.OperatorClient
+	kubeClient     kubernetes.Interface
+	dynamicClient  dynamic.Interface
+	eventRecorder  events.Recorder
+}
+
+const (
+	monitoringControllerName = "VSphereProblemDetectorMonitoringController"
+)
+
+func newMonitoringController(
+	clients *csoclients.Clients,
+	eventRecorder events.Recorder,
+	resyncInterval time.Duration) factory.Controller {
+
+	c := &monitoringController{
+		operatorClient: clients.OperatorClient,
+		kubeClient:     clients.KubeClient,
+		dynamicClient:  clients.DynamicClient,
+		eventRecorder:  eventRecorder,
+	}
+	return factory.New().
+		WithInformers(c.operatorClient.Informer()).
+		WithSync(c.sync).
+		ResyncEvery(time.Minute).
+		WithSyncDegradedOnError(clients.OperatorClient).
+		ToController(monitoringControllerName, eventRecorder.WithComponentSuffix("vsphere-monitoring-controller"))
+}
+
+func (c *monitoringController) sync(ctx context.Context, syncContext factory.SyncContext) error {
+	opSpec, _, _, err := c.operatorClient.GetOperatorState()
+	if apierrors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	if opSpec.ManagementState != operatorapi.Managed {
+		return nil
+	}
+
+	serviceMonitorBytes := generated.MustAsset("vsphere_problem_detector/11_service_monitor.yaml")
+
+	_, err = resourceapply.ApplyServiceMonitor(c.dynamicClient, c.eventRecorder, serviceMonitorBytes)
```

Comment:
Would it make sense move ServiceMonitor into `VSphereProblemDetectorDeploymentController`? There's lot of biolerplate just for these two lines.

Reply:
I was thinking we will need to sync `PrometheusRule` too when I create some alert manager rules and hence it will be better to keep them separate.